src/workerd/api/node/buffer.c++

```diff
@@ -597,5 +606,211 @@ uint32_t BufferUtil::write(
   return writeInto(js, buffer, string, offset, length, getEncoding(encoding));
 }
 
+// ======================================================================================
+// StringDecoder
+
+namespace {
+inline kj::byte getMissingBytes(kj::ArrayPtr<kj::byte> state) {
+  return state[BufferUtil::kMissingBytes];
+}
+
+inline kj::byte getBufferedBytes(kj::ArrayPtr<kj::byte> state) {
+  return state[BufferUtil::kBufferedBytes];
+}
+
+inline kj::byte* getIncompleteCharacterBuffer(kj::ArrayPtr<kj::byte> state) {
+  return state.begin() + BufferUtil::kIncompleteCharactersStart;
+}
+
+inline Encoding getEncoding(kj::ArrayPtr<kj::byte> state) {
+  return static_cast<Encoding>(state[BufferUtil::kEncoding]);
+}
+
+v8::Local<v8::String> getBufferedString(jsg::Lock& js, kj::ArrayPtr<kj::byte> state) {
+  KJ_ASSERT(getBufferedBytes(state) <= BufferUtil::kIncompleteCharactersEnd);
+  auto ret = toStringImpl(js, state,
+                          BufferUtil::kIncompleteCharactersStart,
+                          BufferUtil::kIncompleteCharactersStart + getBufferedBytes(state),
+                          getEncoding(state));
+  state[BufferUtil::kBufferedBytes] = 0;
+  return ret;
+}
+}  // namespace
+
+v8::Local<v8::String> BufferUtil::decode(jsg::Lock& js,
```

Comment:
It's helpful to review a bit about how the implementation works here.

`StringDecoder` is a streaming decoder that ensures that multi-byte characters are correctly handled.  So, for instance, let's suppose I have the utf8 bytes for a euro symbol (0xe2, 0x82, 0xac), but I only get those one at a time... `StringDecoder` will ensure that those are correctly handled over multiple calls to `write(...)`...

```
const sd = new StringDecoder();
let results = '';
results += sd.write(new Uint8Array([0xe2]));  // results.length === 0
results += sd.write(new Uint8Array([0x82]));  // results.length === 0
results += sd.write(new Uint8Array([0xac]));  // results.length === 1
results += sd.end();
```

Internally, the decoder allocates a small 7 byte buffer (the `state`) argument below.

The first four bytes of the `state` are used to hold partial bytes received on the previous write. The fifth byte in `state` is a count of the number of missing bytes we need to complete the character. The sixth byte in `state` is the number of bytes that have been encoded into the first four. So, in our example above, initially the first six bytes of the `state` are `[0x00, 0x00, 0x00, 0x00, 0x00, 0x00]`

After the first call to write above, the state is updated to: `[0xe2, 0x00, 0x00, 0x00, 0x02, 0x01]`

After the second call to write, the state is updated to: `[0xe2, 0x82, 0x00, 0x00, 0x01, 0x02]`

After the third call to write, the pending multibyte character is completed, the state becomes: `[0xe2, 0x82, 0xac, 0x00, 0x00, 0x00]` ... while the bytes are still in state, the buffered bytes and bytes needed are zeroed out. Since the character is completed on that third write, it is included in the returned string.